File: Python/isPowerOfThree.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: isPowerOfThree.py
@Function: 3的幂 数学问题
@Link: https://leetcode-cn.com/problems/power-of-three/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-19
"""


import math

class Solution:
    def isPowerOfThree(self, n: int) -> bool:
        if n <= 0:
            return False

        res = round(math.log(n, 3), 9)
        # python 需要精确到小数点 9 位，才能精确判断小数和它转换后的整数是否相等
        return res - round(res) == 0    # round不加第二个参数，默认保留为整数


# ----------------------------------
if __name__ == "__main__":
    # num = 21
    num = 27

    solution = Solution()
    bool_power_three = solution.isPowerOfThree(num)
    print(f"The solution of this problem is {bool_power_three}")

File: C++/isPowerOfThree.cpp
/**
 * @File    : isPowerOfThree.cpp
 * @Brief   : 3的幂 数学问题 
 * @Link    : https://leetcode-cn.com/problems/power-of-three/
 * @Author  : Wei Li
 * @Date    : 2021-07-19
*/

#include <iostream>
#include <cmath>

class Solution
{
public:
    bool isPowerOfThree(int n)
    {
        // <cmath> std::fmod(x, y) 计算除法运算 x / y 后的浮点形式的余数
        // log_base(3)_(n) 通过换底公式进行计算
        return std::fmod(log10(n) / log10(3), 1) == 0;
    }
};

// ------------------------------
int main(int argc, char **argv)
{
    // int num = 21;
    int num = 27;

    Solution solution;
    auto check_power_three = solution.isPowerOfThree(num);
    std::cout << "The solution of this problem is " << std::boolalpha << check_power_three << std::endl;

    return 0;
}
